coordinator/src/position/mod.rs

```diff
@@ -1 +1,208 @@
+                        )
+                        .await?;
+                        Some(closing_price)
+                    }
+                    false => None,
+                };
+
+                if db::positions::Position::set_open_position_to_closing(
+                    &mut conn,
+                    &to_secp_pk_30(trader_id),
+                    closing_price,
+                )? > 0
+                {
+                    tracing::info!(%trader_id, "Set open position to closing after the dlc channel got force closed.");
+                } else {
+                    tracing::warn!(%trader_id, "Failed to set open position to closing after the dlc channel got force closed.");
+                }
+            }
+            DlcChannelEvent::Closed(_) | DlcChannelEvent::CounterClosed(_) => {
+                let channel = &self.inner.get_dlc_channel_by_reference_id(protocol_id)?;
+                let trader_id = channel.get_counter_party_id();
+
+                let position = match db::positions::Position::get_position_by_trader(
+                    &mut conn,
+                    to_secp_pk_30(trader_id),
+                    vec![PositionState::Closing { closing_price: 0.0 }], /* the closing price doesn't matter here. */
+                )? {
+                    Some(position) => position,
+                    None => return Ok(()),
+                };
+
+                let oracle = self.inner.oracles.first().context("Missing oracle")?;
+                let closing_price = fetch_closing_price_from_attestation(
+                    oracle.host.clone(),
+                    position.expiry_timestamp,
+                )
+                .await?;
+
+                tracing::debug!(
+                    ?position,
+                    trader_id = %trader_id,
+                    "Finalize closing position after force closure",
+                );
+
+                let trader_realized_pnl_sat = {
+                    let trader_position_direction = position.trader_direction;
+
+                    let (initial_margin_long, initial_margin_short) =
+                        match trader_position_direction {
+                            Direction::Long => {
+                                (position.trader_margin, position.coordinator_margin)
+                            }
+                            Direction::Short => {
+                                (position.coordinator_margin, position.trader_margin)
+                            }
+                        };
+
+                    calculate_pnl(
+                        Decimal::from_f32(position.average_entry_price)
+                            .expect("to fit into decimal"),
+                        closing_price,
+                        position.quantity,
+                        trader_position_direction,
+                        initial_margin_long as u64,
+                        initial_margin_short as u64,
+                    )?
+                };
```

Comment:
📣 Surely we should get the PNL from the CET or the settle transaction if we get force-closed. Using this function won't be accurate.

Reply:
We are only interested in the CET as for the settle transaction the position must have been closed.

How would you get the pnl of that position from the CET? Doesn't it include the reserve?

Reply:
> We are only interested in the CET as for the settle transaction the position must have been closed.

That's not true. A settled channel (one without a DLC in it) can be force-closed too.

> How would you get the pnl of that position from the CET? Doesn't it include the reserve?

You would take the payout and subtract the reserve, which is now stored in the `dlc_channels` table.

What you are doing here is an approximation, which maybe is good enough, but I wasn't sure if that's what we wanted to do.

Reply:
> You would take the payout and subtract the reserve, which is now stored in the `dlc_channels` table.
> 
> 
> 
> What you are doing here is an approximation, which maybe is good enough, but I wasn't sure if that's what we wanted to do.

I think that would work as well, but then the pnl would include on-chain fees. That wouldn't be consistent with the pnl calculation of other positions.

I was thinking to calculate the fees per channel differently. But I may be missing something.

Reply:
> I think that would work as well, but then the pnl would include on-chain fees. That wouldn't be consistent with the pnl calculation of other positions.

The CET payouts are not reduced by on-chain fees. The on-chain fees come from the fee reserve. See https://github.com/p2pderivatives/rust-dlc/blob/883ad23f4c6467c90754ccce0b8820db10479f08/dlc/src/lib.rs#L572-L592.